docs/architecture/adr-013-multisigs.md

```diff
@@ -0,0 +1,148 @@
+}
+```
+We will derive N from the length of pubkeys. (For spatial efficiency in encoding)
+
+It will expect to verify an `[]byte` encoded version of the Multisignature.
+(Multisignature is described in the next section)
+The multisignature will be rejected if the bitmap has less than k indices,
+or if any signature at any of the k indices is not a valid signature from
+the kth public key on the message.
+(If more than k signatures are included, all must be valid)
+
+The reason this doesn't use `log_8(n)` bytes per signer is because that heavily optimizes for the case where a very small number of signers are required.
+e.g. for `n` of size `24`, that would only be more space efficient for `k < 3`. 
+This seems less likely, and that it should not be the case optimized for.
+
+#### Weighted threshold signature
+
+The pubkey is the following struct:
+
+```golang
+type WeightedThresholdMultisignatureKey struct {
+	Weights []uint             `json:"weights"`
+	Threshold uint             `json:"threshold"`
+	Pubkeys []crypto.Pubkey    `json:"pubkeys"`
+}
+```
+Weights and Pubkeys must be of the same length.
+Everything else proceeds identically to the K of N multisig, 
+except the multisig fails if the sum of the weights is less than the threshold.
+
+#### Multisignature
+
+The inter-mediate phase of the signatures (as it accrues more signatures) will be the following struct:
+```golang
+type Multisignature struct {
+	BitArray    CryptoBitArray // Documented later
+	Sigs        [][]byte       // []byte instead of crypto.Signature per https://github.com/tendermint/tendermint/issues/1957
+}
+```
+
+It is important to recall that each private key will output a signature on the provided message itself.
+So no signing algorithm ever outputs the multisignature.
+The UI will take a signature, cast into a multisignature, and then keep adding
+new signatures into it, and when done marshal into `[]byte`.
+This will require the following helper methods:
+```golang
+func SigToMultisig(sig Signature, n int)
+func GetIndex(pk crypto.Pubkey, []crypto.Pubkey])
+func AddSignature(sig Signature, index int, multiSig *Multisignature)
+```
+The multisignature will be converted to an `[]byte` using amino.MarshalBinaryBare. \*
+
+#### Bit Array
+We would be using a new implementation of a bitarray. The struct it would be encoded/decoded from is 
+```golang
+type CryptoBitArray struct {
+	ExtraBitsStored  byte      `json:"extra_bits"` // The number of extra bits in elems. 
+	Elems            []byte    `json:"elems"`
+}
+```
+The reason for not using the BitArray currently implemented in `libs/common/bit_array.go`
+is that it is less space efficient, due to a space / time trade-off.
+Evidence for this is outlined in [this issue](https://github.com/tendermint/tendermint/issues/2077). 
+
+In the multisig, we will not be performing arithmetic operations,
+so there is no performance increase with the current implementation,
+and just loss of spatial efficiency.
```

Comment:
Am I the only one who is  not entirely sure what "spatial efficiency" in this context is?

Reply:
I will reword this to "it just takes more space."

Reply:
Cause `[1]uint64` == `[4]byte` in length so CompactBitArray allows for more space efficient bit arrays